React/NextJS routing.md

```diff
@@ -0,0 +1,428 @@
+
+- `query` property will be an empty object at first when accessing the page built with ASG. 
+This is explained above in the `Other information` section, but we want to emphasize this again. To solve it, you can 
+return nothing (null) when `isReady` prop from the `useRouter` returns `false`.
+- Do not use `router.push` if you are handling with the external urls. Use `window.location` instead.
+- The pages navigated via `router.push` are not prefetched, and you need to manually add a `router.prefetch` function call.
+Also, the prefetch works only in the production build.
+- Do not use `withRouter` HOC in React with hooks.
+- At the moment there is no clean NextJS solution to handle the cancellation of the routes and their transition, but you
+can check this [Route cancellation dirty solution](https://github.com/vercel/next.js/discussions/32231#discussioncomment-2421565)
+
+
+## Link
+
+The `Link` component is used for the internal client-side links. It renders the content differently based on the context.
+Altough for plain text it will render an `a` element, always have the `a` element as the outermost element in `Link` children. 
+
+```tsx
+// all render <a href="/">Home</a>
+<Link href="/">Home</Link> // notice only plain text; try to avoid this and always put the <a> tag
+<Link href="/"><a>Home</a></Link>
+<Link href="/" passHref><CustomLink>Home</CustomLink></Link> // notice passHref here
+
+// notice forwardRef here
+const CustomLink = forwardRef(({ children, ...rest }, ref) => {
+	return (
+		<a ref={ref} {...rest}>
+			{children}
+		</a>
+	);
+});
+```
+
+Prefer to use the NextJS `Link` component outside the implementation:
+```tsx
+// BAD
+const Card = (...) => {
+    return (
+        <Link href={href}>{...}</Link>
+    );
+}
+
+<Card href="/company/123" {...} />
+
+
+// GOOD
+const Card = (...) => {
+    return (
+        ...
+    );
+}
+
+<Link href="/company/123">
+    <a>
+        <Card>{...}</Card>
+    </a>
+</Link>
+```
+
+By default, the links created with the `Link` component will prefetch the pages. `Link` uses 
+[`Intersection Observer API`](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API)
+to detect when to load with a `rootMargin` of `200px`. At the moment observer values can not be changed (checked in NextJS 12.1.4).
+You can turn off this feature if needed by setting `false` for the prefetch` prop.
+
+For the API reference, please check the official docs for the [`Link` component](https://nextjs.org/docs/api-reference/next/link).
+
+
+## Base path
+
+You can change the base path if your application needs to be under a sub-path of a domain.
+To do that, change the `basePath` property in the `next.config.js` file.
+
+
+## 404 pages
+
+Routing to non-existing pages is working out of the box. 
+Only if there is a `SSG` or a `SSR` then you need to handle it manually. To implement that, you should return: 
+```
+notFound: true
```

Comment:
Could you add the whole example with `getSetverSideProps`

Reply:
Sure